src/components/visual-programming/CodeGraph.vue

```diff
@@ -696,44 +623,59 @@ async function logCode() {
     "\n" +
     NoiseFunction +
     "\n\nfn sampleObject(input2: vec2f) -> vec3f {\n" +
-    (await orderedCode(allNodes, visited, "\t"));
+    orderedCode(returnNodes, "\t");
 
   fullCode += "\n}";
   emit("update", fullCode);
 }
 
-async function orderedCode(
-  allNodes: Nodes[],
-  visited: string[],
-  indent: string = ""
-) {
-  if (allNodes.length <= 0) return "";
-  const graph = structures(editor);
+function orderedCode(startNodes: Nodes[], indent: string = "") {
+  const allConnections = inputConnections();
   let fullCode = "";
-  let nodeQueue = allNodes;
-
-  while (nodeQueue.length > 0) {
-    const node = nodeQueue.shift();
-    if (!node) break;
-    if (visited.includes(node.id)) continue;
-    if (node instanceof NothingNode) {
-      await editor.removeNode(node.id);
-      continue;
+
+  // Permanent mark
+  const nodeOutput = new Map<NodeId, NodeData>();
+  const temporaryMark = new Set<NodeId>();
+
+  // https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
+  function visit(node: Nodes) {
+    if (nodeOutput.has(node.id)) return;
+    if (temporaryMark.has(node.id)) {
+      showError("Graph is cyclic!");
+      return;
     }
-    const incomers = graph.incomers(node.id).nodes();
-    // Check if any incomers of current node weren't visited yet and ignore this node for now if so
-    if (
-      incomers.some((inc) => !visited.includes(inc.id)) ||
-      (node instanceof ReturnNode && nodeQueue.length > 0)
-    ) {
-      nodeQueue.push(node);
-      continue;
+
+    temporaryMark.add(node.id);
+
+    const connections = allConnections.get(node.id) ?? {};
+    for (const connection of Object.values(connections)) {
+      const child = editor.getNode(connection.source);
+      assert(child, "invalid connection");
+      visit(child);
     }
 
-    visited.push(node.id);
-    fullCode += await getNodesCode(node, visited, graph, indent);
+    const inputKeys = Object.keys(node.inputs);
+
+    const inputData: Partial<NodeData> = Object.fromEntries(
+      inputKeys.map((v) => {
+        const connection = connections[v];
+        if (connection) {
+          return [
+            v,
+            nodeOutput.get(connection.source)?.[connection.sourceOutput],
+          ];
+        } else {
+          return [v, undefined];
+        }
+      })
+    );
```

Comment:
This construction right here is the most complicated bit I'd say.

If you don't immediately understand it, here's what it does:
1. Take all the names of inputs of a node (like x, y, z for a join node)
2. Follow their connections
3. Read the relevant output
4. And assemble that into an object like `{ x: someOutput, y: undefined, z: someOutput }`